xmodule/modulestore/__init__.py

```diff
@@ -313,13 +319,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
         if signal_handler and bulk_ops_record.has_publish_item:
             signal_handler.send("pre_publish", course_key=course_id)
 
+    def create_catalog_data_for_signal(self, course_id):
+        """
+        Creates a CourseCatalogData object from a course id by fetching from the module store
+
+        Arguments:
+            course_id (CourseKey): id of the course to process
+
+        Returns:
+            CourseCatalogData
+        """
+        # Most of this is copied from CourseOverview. We cannot import it
+        # because it would lead to circular imports
+        course_from_store = self.get_course(course_id)
```

Comment:
Where is `self.get_course` defined?

Reply:
Turns out this is _probably_ defined on all descendents; we should consider adding an abstract method here. One wrinkle: Some subclasses call it course_id, some call it course_key. Maybe make it a positional-only param, if we can?

Reply:
Proposal from discussion: Rather than defining `get_course` and `get_item` as abstract methods on this mixin...

1. Move the code outside of this modulestore class and into a handler for the existing `course_published` signal.
2. Update the code to call `modulestore()` to _get_ a modulestore instance and call its `get_course` method.
3. Call `CourseDetails.fetch_about_attribute` in order to get the effort information. (Might not create a circular dependency this way...)

@ormsbee Is there anything we need to worry about here with regard to transactionality? I don't know how Mongo interacts with our implicit transactions, and whether we should be 1) making sure we only build and send this event after the transaction has successfully completed (so as not to send "ghost" updates in the case of tx rollbacks), and/or 2) whether there are risks of getting the course and effort item separately (consistency issues).

Reply:
(I've pushed a new commit, 95d3fdd, which implements this proposal.)

Reply:
There are four ModuleStore classes that are instantiated, and you can rely on `get_course` to exist in them. Please don't add an abstract method. ModuleStore has a pile of inheritance soup, but we're slowly making progress killing off the worst pieces of it via BD-13 and some new Old-Mongo-DEPR work slated to start soon.

The approach you took to move it to a handler in contentstore is fine.

In terms of transactionality–as long as you're doing it from the handler after catching a publish signal, you're probably fine. There are situations in which it can get messed up, but that affects just about everything, and I don't think you should try to solve for it in the scope of these efforts.

The one thing you will encounter a lot though is that you're going to get these signals for both draft and publish branch changes. Since you only care about things that have been published to the LMS, you might want to do something like:

https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/outlines.py#L336-L338

Also, there is a weird thing that happens to UsageKeys for changed items in a publish when you ask for the published branch in Studio. It probably doesn't affect you (since you're not really reading any UsageKeys directly), but just in case, please read: https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/outlines.py#L25-L43

Reply:
- I've moved the signal-send to be after transaction commit, although I wonder if that deferral should actually be done by the signal-event translation layer (that we haven't written yet).
- I've used the `branch_setting` context manger you pointed out. I don't really understand all the details of branches, but I think this along with the ` course_key=course_key.for_branch(None)` should give the "published only" view of the course.
- I definitely don't know much about usage keys, but it looks like it won't affect this.